keps/sig-scheduling/5055-device-taints-and-tolerations/README.md

```diff
@@ -0,0 +1,982 @@
+used for scheduling new pods. In addition, pods already running with access to
+a tainted device can be stopped automatically. Cluster administrators can do
+the same by creating a
+[ResourceSliceOverride](../5027-dra-admin-controlled-device-attributes) with a
+taint.
+
+Users can decide to ignore specific taints by adding tolerations to their
+ResourceClaim.
+
+## Motivation
+
+### Goals
+
+- Enable taking devices offline for maintenance while still allowing tests pods
+  to request and use those devices. Being able to do this one device at a time
+  minimizes service level disruption.
+
+- Enable users to decide whether they want to keep running a workload in a degraded
+  mode while a device is unhealthy or prefer to get pods rescheduled.
+
+### Non-Goals
+
+- Not part of the plan for alpha: developing a kubectl command for tainting devices.
+  This may be reconsidered.
+
+## Proposal
+
+### User Stories
+
+#### Degraded Devices
+
+A driver itself can detect problems which may or not be tolerable for
+workloads, like degraded performance due to overheating. Removing such devices
+from the ResourceSlice would unconditionally prevent using them for new
+pods. Instead, publishing with a taint informs users about this degradation and
+leaves them the choice whether the device is still usable enough to run pods.
+It also automates stopping pods which don't tolerate such a degradation.
+
+#### External Health Monitoring
+
+As cluster admin, I am deploying a vendor-provided DRA driver together with a
+separate monitoring component for hardware aspects that are not available or
+not supported by that DRA driver. When that component detects problems, it can
+check its policy configuration and decide to take devices offline by creating
+a ResourceSliceOverride with a taint for affected devices.
+
+### Risks and Mitigations
+
+A device can be identified by its names (`<driver name>/<pool name>/<device
+name>`) and/or by its attributes (for example, a unique ID). It was a conscious
+decision for core DRA to not require that the name is tied to one particular
+hardware instance to support hot-swapping. Admins might favor using the names
+whereas health monitoring might prefer to be specific and use a vendor-defined
+unique ID. Both are supported, which creates additional complexity.
+
+Without a kubectl extension similar to `kubectl taint nodes` the user
+experience for admins will be a bit challenging. They need to decide how to
+identify the device (by name or with a CEL expression), manually create a
+ResourceSliceOverride with a unique name, then remember to remove that
+ResourceSliceOverride again. For beta, support in `kubectl` for common
+operations may be needed.
+
+## Design Details
+
+The feature is following the approach and APIs taken for node taints and
+applies them to devices. A new controller watches tainted devices and deletes
+pods using them unless they tolerate the device taint, similar to the
```

Comment:
Operators like training operator (https://github.com/kubeflow/training-operator) can spawn more pods when pods are deleted, will pod deletion cause pending pods in the cluster?

Reply:
Pods can also have finalizers set by different operators

Reply:
A finalizer doesn't prevent stopping the containers. As soon as the kubelet sees a pod with DeletionTimestamp, it will stop it. That's sufficient for this KEP.

The pod object then remains in the API server with a final "stopped and deleted" state. That in turn is sufficient for DRA to deallocated claims, so this also doesn't block devices.

Reply:
Thanks. If an operator pod spawns the next pod referencing the same claim, will such a claim be deallocated?

Reply:
_Hi, I'm one of this KEP authors._

In this design, Taints will be defined per device in ResourceSlice(Override). Tolerations will be declared per device in ResourClaim.

> If an operator pod spawns the next pod referencing the same claim, will such a claim be deallocated?

Thus, if multiple pods shared ResourceClaims and some allocated devices in them are tainted, and the claim can not tolerate the taints, yes, all the pods will be deleted and the resource claim will be unallocated.

Reply:
> A finalizer doesn't prevent stopping the containers. ...

Added also as text in the KEP.

Reply:
Thanks @everpeace. this may be good for workloads that have all-or-nothing semantics (e.g., AI workloads); on the downside, independent workloads consuming a single claim will have a hit even if one of the independent pods crashes. is my understanding correct?

Reply:
> independent workloads consuming a single claim will have a hit even if one of the independent pods crashes. is my understanding correct?

Yes, you're correct. But, I don't think it's downside.

IF multiple pods shared the same ResourceClaim and the device allocated the devices to the claim was tainted(effect=NoExecute), I think it's natural to delete all the pods using the ResourceClaim because the claim can not tolerate the taints(effect=NoExecute).

Reply:
> independent workloads consuming a single claim will have a hit even if one of the independent pods crashes

A pod crashing seems unrelated to me. It doesn't cause the claim to get tainted, so other pods sharing the claim are not affected.